pubsub/event.go

```diff
@@ -0,0 +1,307 @@
+}
+
+// Star creates a Field containing the Kleene star `*` for pattern subscription,
+// and chains it on to the Event.
+func (e Event) Star() Event {
+	e.assertPattern()
+	e.fields = append(e.fields, Field{valid: true, value: "*", pattern: patternStar})
+	return e
+}
+
+// Placeholder creates a field containing a `?` for a placeholder in Redis patterns,
+// and chains it on to the event.
+func (e Event) Placeholder() Event {
+	e.assertPattern()
+	e.fields = append(e.fields, Field{valid: true, value: "?", pattern: patternPlaceholder})
+	return e
+}
+
+// Alternatives creates a field with the alts wrapped in brackets, to match
+// one of them in a Redis pattern, and chains it on to the event.
+func (e Event) Alternatives(alts string) Event {
+	e.assertPattern()
+	e.fields = append(e.fields, Field{
+		valid:   true,
+		value:   "[" + alts + "]",
+		pattern: patternAlts,
+	})
+	return e
+}
+
+// assertPattern panics if the event is not a Pattern type.
+func (e Event) assertPattern() {
+	if e.kind != PatternEvent {
+		panic("That operation is only valid on pattern events created with NewPattern()")
+	}
+}
+
+// Len returns the number of fields contained in the event.
+func (e Event) Len() int { return len(e.fields) }
+
+// Get returns the value of a field at index `i` within the event. If the
+// field does not exist, an empty struct will be returned.
+func (e Event) Get(i int) Field {
+	if len(e.fields) <= i {
+		return Field{valid: false}
+	}
+
+	return e.fields[i]
+}
+
+// Find looks up a field value by its alias. This is most useful in pattern
+// subscriptions where might use Find to look up a parameterized property.
+// If the alias does not exist, an empty struct will be returned.
+func (e Event) Find(alias string) Field {
+	for _, field := range e.fields {
+		if field.alias == alias {
+			return field
+		}
+	}
+
+	return Field{valid: false}
+}
+
+// Name returns name of the event, formed by a concatenation of all the
+// event fields.
+func (e Event) Name() string {
+	strs := make([]string, len(e.fields))
+	for i, field := range e.fields {
+		strs[i] = field.value
+	}
+
+	return strings.Join(strs, "")
+}
+
+// Type returns the type of the event (either a PlainEvent or a PatternEvent).
+func (e Event) Type() EventType {
+	return e.kind
+}
+
+func (e Event) concat(other Event) Event {
+	e.fields = append(e.fields, other.fields...)
+	return e
+}
+
+func (e Event) slice(start, end int) Event {
+	e.fields = e.fields[start:end]
+	return e
+}
+
+func (e Event) isZero() bool {
+	return e.kind == 0
+}
+
+// applyFields attempts to convert the values from a string or byte slice into
+// a Field. It panics if a value is none of the above.
+func applyFields(event Event, values []interface{}) Event {
+	for _, v := range values {
+		switch t := v.(type) {
+		case string:
+			event = event.String(t)
+		case []byte:
+			event = event.String(string(t))
+		default:
+			panic(fmt.Sprintf("Expected string or field when creating an event, got %T", v))
```

Comment:
`applyFields` is used as a shortcut when you just want to make a subscriber to an event, like `NewEvent("something")`. A panic would be caused by:
 - the developer screwing up and putting the wrong type into NewEvent
 - the developer screwing up and passing an `interface{}` from some external source in without knowing/checking its type

In both cases it's a programmer error, so I think a panic is appropriate